A36-xds-for-servers.md

```diff
@@ -0,0 +1,385 @@
+`XdsServingStatusListener` will be defined as:
+
+```java
+package io.grpc.xds;
+
+public interface XdsServingStatusListener {
+  void onServing();
+  void onNotServing(Throwable t);
+}
+```
+
+It is an interface instead of an abstract class as additional methods are not
+expected to be added before Java 8 language features are permitted in the code
+base. If this proves incorrect, an additional interface can be added.
+
+If not specified, a default implementation of `XdsServingStatusListener` will be
+used. It will log the exception and log calls to `onServing()` following a call
+to `onNotServing()` at WARNING level.
+
+In order to allow transport-specific configuration, the `XdsServerBuilder` will
+have a `@ExperimentalApi ServerBuilder transportBuilder()` method whose return
+value can be cast to `NettyServerBuilder` for experimental API configuration.
+
+`NettyServerBuilder` will add an internal method `eagAttributes(Attributes)`. It
+will plumb those attributes to `GrpcHttp2ConnectionHandler.getEagAttributes()`
+as implemented by `NettyServerHandler`, which currently just returns
+`Attributes.EMPTY`. `XdsServerBuilder` will then specify `eagAttributes` to
+inject credential information for the `XdsServerCredential`, similar to
+client-side, although with the added step of needing to process the
+FilterChainMatch for the specific connection.
+
+
+### Go
+
+Create an `xds.GRPCServer` struct that would internally contain an unexported
+`grpc.Server`. It would inject its own `StreamServerInterceptor` and
+`UnaryServerInterceptor`s into the `grpc.Server`. The interceptors would be
+controlled by an `atomic.Value` or mutex-protected field that would update with
+the configuration. 
+
+`GRPCServer.Serve()` takes a `net.Listener` instead of a `host:port` string to
+listen on, so as to be consistent with the `Serve()` method on the
+`grpc.Server`. The implementation expects the `Addr()` method on the passed in
+`net.Listener` to return an address of type `net.TCPAddr`. It then creates an
+XdsClient and observes the `Addr()` to create a watch.  Before
+`Serve(net.Listener)` returns, it will cancel the watch registered on the
+XdsClient. As part of `Stop()/GracefulStop()`, the XdsClient is shut down. Note
+that configuration is nominally per-address, but there is only one set of
+interceptors, so the interceptors will need to look up the per-address
+configuration for each RPC.
+
+Service registration is done on a method in the generated code which
+accepts a `grpc.ServiceRegistrar` interface. Both `grpc.Server` and 
+`xds.GRPCServer` implement this interface and therefore the latter can be passed to
+service registration methods just like the former.
+
+
+```
+// instead of
+s := grpc.NewServer()
+pb.RegisterGreeterServer(s, &server{})
+// it'd be
+s := xds.NewGRPCServer()
+pb.RegisterGreeterServer(s, &server{})
+```
+
+
+Package `credentials/xds` exports a `NewServerCredentials()` function which
```

Comment:
@easwars, will the `NewServerCredentials()` have an argument for 'fallback' credentials if the xds server does not provide credentials?